keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,916 @@
+- c) client specifies a timeout in the request URI `{host/path}?timeout=10s`: yes, or no
+- d) the state of the `ResponseWriter` object: 
+  - the request handler neither does a `Write`, not `FlushError` before timeout occurs
+  - the request handler does a `Write`, but not `FlushError` before timeout occurs
+  - the request handler does a `Write`, followed by a `FlushError` before timeout occurs
+
+
+Similarly, we should assert on the following, to cover the expectations on both the client and the server:
+
+- server: 
+  - expected error from `Write` and `FlushError` method from the `ResponseWriter` object after timeout
+  - that the request handler terminates.
+- client:
+  - expected  `response` and `err` from `client.Do(req)`
+  - expected error while reading the `Body` of the `http.Response` sent by the server.
+
+
+These tests will very closely resemble what we would see in production 
+environment since these tests exercise the entire handler chain of the
+kube-apiserver
+
+> We can also use the tests where `PerHandlerReadWriteTimeout == false` to document
+> what the client sees today with the timeout filter enabled, and compare with the 
+> results from tests where `PerHandlerReadWriteTimeout == true` to understand how 
+> differently and the client is impacted with this new feature enabled.
+
+
+##### Integration tests
+
+Two tests:
+
+- Scenario A: `PerHandlerReadWriteTimeout` is disabled, a request times out
+- Scenario A: `PerHandlerReadWriteTimeout` is enabled, a request times out
+
+
+##### e2e tests
+
+Similar to Integration tests
+
+### Graduation Criteria
+
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- All new tests enumerated in "Test Plan" are implemented
+
+#### Beta
+
+TBD
+
+#### GA
+
+TBD
+
+#### Deprecation
+
+
+### Upgrade / Downgrade Strategy
+
+
+### Version Skew Strategy
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [ ] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: PerHandlerReadWriteTimeout
+  - Components depending on the feature gate:
+	  - kube-apiserver
+
+###### Does enabling the feature change any default behavior?
+
+An API client will no longer receive the `504 GatewayTimeout` HTTP status code
+when the server times out its request. Instead, the client will receive an error 
+
+NOTE: the client can not rely on the receipt of the `504 GatewayTimeout` HTTP status
+code as it is not a deterministic error path on the server side.
+
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, through the feature gate 
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+No additional considerations, except the feature is enabled.
+
+###### Are there any tests for feature enablement/disablement?
+
+There will be both unit tests and integration tests that asserts on:
```

Comment:
This feature is purely in-memory.

Enablement/disablement tests are supposed to test the "process of disablement/enablement" (i.e. switching on-off the feature gate during the test).
In case of this feature enablement/disablement is no different that just starting kube-apiserver with/without the features.
So those aren't strictly needed for features that don't persist any state.